[12.0] (shop)?invader_payment_{sips,stripe}, (shop)?invader_payment

rousseldenis opened the pull request:
This is a first functional version.

A great simplification was done to make specific acquirers implementation as simple as possible and as diconnected as possible from too complex code.

One thing to note: to do that, we removed the dependency on payment-gateway modules.

rousseldenis committed:
[12.0][MIG] shopinvader_payment, shopinvader_payment_stripe

rousseldenis commented:
@lmignon @sbidoul @sebastienbeau

lmignon commented on shopinvader_payment_stripe/services/payment_stripe.py:
```diff
@@ -0,0 +1,273 @@
+        target,
+        payment_mode=False,
+        stripe_payment_method_id=False,
+        stripe_payment_intent_id=False,
```
@rousseldenis IMO confirm_payment should takes **kwargs as params since the schema is defined by cerberus. In this way, it's possible to use the same implementation for others kind of target that could requires additional parameters.

lmignon commented on shopinvader_payment_stripe/services/payment_stripe.py:
```diff
@@ -0,0 +1,273 @@
+        :return:
+        """
+        transaction_obj = self.env["payment.transaction"]
+        target = self._find_target()
```
@rousseldenis  `_find_target (target, **params)`

lmignon commented on shopinvader_payment/models/shopinvader_payable.py:
```diff
@@ -0,0 +1,119 @@
+        """
+        raise NotImplementedError
+
+    def _get_target_provider(self, target):
```
@rousseldenis This method is never used

lmignon commented on shopinvader_payment_stripe/services/payment_stripe.py:
```diff
@@ -0,0 +1,273 @@
+            "error": {"type": "string"},
+        }
+
+    def _find_target(self):
```
@rousseldenis  `def _find_target(target, **params)`

```python
if target == 'current_cart':
    ...
else:
    raise NotImplementedError()
```

lmignon commented on shopinvader_payment_stripe/services/payment_stripe.py:
```diff
@@ -0,0 +1,273 @@
+        payment_mode=False,
+        stripe_payment_method_id=False,
+        stripe_payment_intent_id=False,
+    ):
```
@rousseldenis 
```python
stripe_payment_intent_id=False,
**params
```

In this way this method can be overridden to support additional parameters defined into the Cerberus schema

sbidoul committed:
[IMP] Refactor general directions
The goal is to put a generic REST services layer that could be used to pay payable models, provides the implementation for some aquirer and build the shopinvader specificities on top of this generic layer.

sbidoul commented on invader_payment/models/account_payment_mode.py:
```diff
@@ -0,0 +1,15 @@
+        comodel_name="payment.acquirer", string="Payment Acquirer"
+    )
+    # TODO why do we need this field?
+    provider = fields.Selection(related="payment_acquirer_id.provider")
```
Don't forget to add the view.

sbidoul committed:
[ADD] setup.py

sbidoul commented on invader_payment_stripe/services/payment_stripe.py:
```diff
@@ -0,0 +1,273 @@
+class PaymentServiceStripe(Component):
+
+    _name = "payment.service.stripe"
+    _inherit = "base.shopinvader.service"
```
invader_stripe must not depend on shopinvader.

sbidoul committed:
[IMP] Attempt to clear cart after payment

sbidoul committed:
[IMP] Simplify session code

rousseldenis committed:
[FIX] methods

sbidoul committed:
[FIX] Hack to keep last_sale in store_cache

sbidoul committed:
[IMP] cleanup stripe modules a bit

Plus, add missing shopinvader_payment_stripe that
exposes it under the /shopinvader route.

sbidoul committed:
gitignore

sbidoul committed:
[ADD] invader_payment_sips

This required some refactoring of invader_payment.

lmignon committed:
[FIX] trailing whitespaces and end of files

lmignon commented on shopinvader_payment/models/sale_order.py:
```diff
@@ -0,0 +1,32 @@
+        return vals
+
+    def _invader_payment_start(self, transaction, payment_mode_id):
+        self.write({"payment_mode_id": payment_mode_id.id})
```
@rousseldenis @sbidoul We must play the onchange otherwise the workflow is not set on the SO
```suggestion
        self.ensure_one()
        vals = {"payment_mode_id": payment_mode_id.id}
        newvals = self.play_onchanges(vals, ["payment_mode_id"])
        vals.update(newvals)
        self.write(vals)
```

With this change, this addon should depends on 'onchange_helper"

sbidoul replied:
done in 35c4e3a

sbidoul committed:
[FIX] Add missing external dependencies

lmignon committed:
[IMP] Add bank transfer payment

sbidoul committed:
[FIX] Play onchanges after settings payment mode on sale order

so sale_automatic_workflow_payment_mode can
assign the workflow to the SO

sebastienbeau reviewed (approved):
I like this new implementation. I thinks it's really better to have a payment service with target.

sebastienbeau commented:
I am going to work on writting test and do a PR for proposing a serie of test

sbidoul commented:
@sebastienbeau thanks! that was on my TODO but will not have time before several weeks.

At this stage, the only think I don't like in this version is `_invader_get_payment_success_reponse_data` which is used to inject the session / store cache.
That's why I started thinking about https://github.com/shopinvader/odoo-shopinvader/pull/371/files but it lead nowhere so far because I got lost between components and models.

Ideally we should be able to manipulate the session from anywhere (component or model), so we could reset the session and set last_sale in `payable._confirm_cart`.

simahawk commented:
Guys, it seems that simple wire/bank transfer payment is not supported. With only `shopinvader_payment` installed there's no endpoint to handle the payment:

```
2019-09-03 12:25:25,893 1 WARNING odoodb odoo.addons.base_rest.components.service: Method add_payment is not a public method of service shopinvader.cart.service 
2019-09-03 12:25:25,901 1 INFO odoodb werkzeug: 172.18.0.3 - - [03/Sep/2019 12:25:25] "POST /shopinvader/cart/add_payment HTTP/1.1" 404 - 11 0.011 0.032
```
Basically the `_add_payment` hook is not implemented on the cart, as it is for quotations in `shopinvader_quotation_payment`.
AFAIS we could easily fix this by adding a similar module called `shopinvader_wire_transfer` w/ the same features more or less.

I could do it but I don't want to overlap w/ your work in case you are already addressing it. What do you think?

BTW: "bank transferT" is wrong, there's one T more :wink: 

EDIT: or simply add the handle into shopinvader_payment.

lmignon commented:
@simahawk We've implemented this kind of payment for odoo9 https://github.com/acsone/odoo-shopinvader-payment/tree/9.0-slg_master
The code should be easy to forward port since it's based on this PR (back ported in 9.0 :-/)

simahawk commented:
@lmignon ahhh great, I'll port them if no one is on it then:

https://github.com/acsone/odoo-shopinvader-payment/tree/9.0-slg_master/invader_payment_bank_transfer
https://github.com/acsone/odoo-shopinvader-payment/tree/9.0-slg_master/shopinvader_payment_bank_transfer

lmignon commented:
@simahawk Great!
I don't know if @thibaultrey has already put this payment method into the generic template but it's very easy to add....

Create a file for the new snippet into  app/views/snippets/payment_bank_transfer.liquid with the following content
```liquid
<div class="row justify-content-center">
	<div class="col-12 col-md-8 col-lg-12 col-xl-8	 d-flex">
		<div class="text-primary pr-2 h4">
			<i class="fas fa-long-arrow-alt-right"></i>
		</div>
		<div>
			<b class="h5">
				{{'payment_method_bank_transfer_title' | translate }}
			</b>
			{% if payment_method.description and payment_method.description != false %}
        <div class="small">
          {{payment_method.description}}
        </div>
      {% endif %}
		</div>
	</div>
	<div class="col-12 col-sm-5 col-md-4 col-lg-5 col-xl-4 text-center">
		<form method="POST" action="{{action}}">
	    <input type="hidden" name="invader_success_url" value="{% path_to 'cart_end' %}" />
	    <input type="hidden" name="invader_error_url" value="{% path_to page %}" />
	    <input type="hidden" name="payment_mode[id]" value="{{payment_method.id }}">
		<button  id="button-pay-bank-transfer" type="submit" class="btn btn-primary btn-sm  btn-block"
			data-paymentmode="{{payment_method.id }}"
        	data-successurl="{% path_to cart_end %}"
	        data-cancelurl="{% path_to page %}"
        	data-target="current_cart">
	      {{'payment_method_bank_transfer_btn' | translate }}
	    </button>
	  </form>
	</div>
</div>
{% capture js_bank_transfer %}
	<script defer="true">
		'use strict';
		window.addEventListener("load", function() {
		  var button = document.getElementById("button-pay-bank-transfer");
		  button.addEventListener('click', function(event){
			event.preventDefault();
			$('.main-loader').removeClass('d-none');
			fetch('/invader/payment_bank_transfer/add_payment', {
			  credentials: "include", // default on modern browser but required but for some old browsers
			  method: 'POST',
			  headers: {
				'Content-Type': 'application/json'
			  },
			  body: JSON.stringify({
				target: 'current_cart',
				payment_mode: button.dataset.paymentmode,
			  })
			}).then(function(result) {
			  if(result.ok) {
				document.location = button.dataset.successurl;
			  }
			  else {
				modal_error('show');
			  }
			}).catch(function(error) {
			  modal_error('show', error.message);
			});
		  });
		}, false);
  </script>
{% endcapture %}
{% assign js_footer = js_footer | append: js_bank_transfer %}
```

And modify the payment panel into app/views/snippets/block_cart_payment.liquid to include this new snippet...
```liquid
<div id="payments-panel" {% if params.canceled_payment == null %}class=" d-none"{% endif %}>
      {% for payment_method in store.cart.payment.available_methods.items %}
        <div class="container card card-body mb-3" >
          {% case payment_method.code %}
         ....

          {% when 'credit_card_sips' %}
            {% include 'payment_sips' with payment_method,
              amount: store.cart.amount.total,
              action: '/invader/payment_sips/prepare_payment' %}
          {% when 'bank_transfer' %}
            {% include 'payment_bank_transfer' with payment_method,
              amount: store.cart.amount.total,
              action: '/payment_bank_transfer/add_payment' %}

           ....

          {% else %}
            Indisponible
          {% endcase %}
        </div>
      {% endfor%}
    </div>

```

simahawk commented:
@lmignon thanks for the pointer... Nope, they were not ready in std template AFAIS. I've done them already before discovering that the final payment was not working :sweat_smile: 
Maybe later I'll open a PR to the locomotive template.
BTW  it would be nice to retrieve the snippet automatically w/out having to add an IF/case block... but this is another story and off-topic here :stuck_out_tongue_winking_eye:

simahawk committed:
Bank transfer: migrate to 12.0

sebastienbeau committed:
[IMP] Add basic test on stripe

sebastienbeau commented:
Testing PR here https://github.com/acsone/odoo-shopinvader-payment/pull/3 more to come

sbidoul committed:
[IMP] Clarify invader_payable interfaces

simahawk committed:
Bank transfer: add readme

simahawk committed:
Fix invader_payment: restore base component

Without this all payment handlers are broken
as they rely on `invader.payment` generic service.

simahawk committed:
Linting

sebastienbeau commented:
After talking with @sbidoul I am in favors of creating transaction (in pending state) for bank transfert (or checkmo is will be the same acquire type just a different payment mode).

Indeed with that logic we always have a transaction (like odoo), in long terme
- we may remove payment_mode on sale order (transaction will already port the information)
- we can imagine to have a module that confirm the "transfert transaction" from the bank statement and so confirm the sale order
- we can handle multi-payment (I have customer case where the customer paid by transfert but later due to change in the order, the difference is paid by credit card, right now is complexe because I have to mix transaction and move line to know if the sale is paid and ready to be ship)

sebastienbeau commented on invader_payment_bank_transfer/services/payment_bank_transfer.py:
```diff
@@ -0,0 +1,46 @@
+    _description = "REST Services for bank transfer payments"
+
+    def _validator_add_payment(self):
+        schema = {"payment_mode": {"type": "string"}}
```
from odoo.addons.base_rest.components.service import to_int

I have a question, why the type is string? we can use 
"id": {"coerce": to_int, "type": "integer"} 

this will avoid to convert it latter as a integer :

sbidoul replied:
Sounds good to me. I don't remember why we kept string. @rousseldenis do you?

sebastienbeau commented on invader_payment_bank_transfer/services/payment_bank_transfer.py:
```diff
@@ -0,0 +1,46 @@
+            usage="invader.payment"
+        )._invader_find_payable_from_target(target, **params)
+        payment_mode = self.env["account.payment.mode"].browse(
+            int(payment_mode)
```
int convertion can be done before

sebastienbeau commented on invader_payment_stripe/services/payment_stripe.py:
```diff
@@ -0,0 +1,253 @@
+            {
+                # payment_mode cannot be integer here, because
+                # it can be an empty string from the form
+                "payment_mode": {"type": "string"},
```
same here convert to_int

sebastienbeau commented on invader_payment_sips/services/payment_sips.py:
```diff
@@ -0,0 +1,268 @@
+
+    def _validator_prepare_payment(self):
+        schema = {
+            "payment_mode": {"type": "string"},
```
same here convert to_int

sebastienbeau committed:
[FIX] fix requirements

sebastienbeau commented on shopinvader_payment/services/invader_payment_service.py:
```diff
@@ -0,0 +1,42 @@
+    def _invader_find_payable_from_target(self, target, **params):
+        if target == "current_cart":
+            return self.component(usage="cart")._get()
+        return super()._invader_find_payable_from_target(target, **params)
```
Maybe we should raise an error here in case that the cart do not exist. To avoid weird issue latter?
@sbidoul @lmignon @rousseldenis

sbidoul replied:
Raising an error sounds good to me. UserError?

sebastienbeau committed:
[FIX] fix missing dependency

sebastienbeau commented:
In the code I see that something we use "payment_mode" and something "payment_method".
Indeed before it was called "payment_method" and now odoo use "account_payment_mode".

As this new version have breaking change, I think it's important to be consistent.
So what we choose payment_mode or payment_method?

Maybe payment_mode to stick to odoo name ?
@lmignon @sbidoul

sbidoul commented:
@sebastienbeau about payment_mode vs payment_method, yes, a clarification is welcome. We must distinguish the OCA `account.payment.mode` (=> variable name `payment_mode`) from the Shopinvader `shopinvader.payment` which is declared on the backend as `payment_method_ids`. I suggest to use `shopinvader_payment_method` as variable name for that.

sebastienbeau commented:
Ok I will propose a PR for this.
I think we can merge this first version of test 
https://github.com/acsone/odoo-shopinvader-payment/pull/3
I will be easier to add other test based on this existing test_invader_payment module

lmignon committed:
[IMP] Improve modularity and reduce interdependencies by making the transaction state changes bubble as event to payable

lmignon committed:
[ADD] setup.py

lmignon commented:
>     * we may remove payment_mode on sale order (transaction will already port the information)

@sebastienbeau Don't forget that the paymen_mode is used to link a sale_automatic_workflow to the SO....

sebastienbeau committed:
[REF] refactor bank_transfer payment

Rename payment bank_transfer into payment_manual
Add acquired for check
Move demo data
Create a transaction in pending state for all manual payment

sebastienbeau committed:
[FIX] fix flake8

sebastienbeau commented:
> > ```
> > * we may remove payment_mode on sale order (transaction will already port the information)
> > ```
> 
> @sebastienbeau Don't forget that the paymen_mode is used to link a sale_automatic_workflow to the SO....

Yes I know, maybe we can set it in a different way. In e-commerce case you always want to generate the invoice automatically after shipping (whatever is the payment method) so maybe we can set a generic workflow per backend and then the logic of order validation can be based on the transaction level and not anymore on the workflow.
But this is not the priority for now

sebastienbeau committed:
[MIG] migrate module shopinvader_quotation_payment

Migrate the module shopinvader_quotation_payment
Add module test_shopinvader_payment
Improve code and fix pep8

sebastienbeau committed:
[IMP] refactor and add test

Fix: add _invader_restrict_payment_mode_ids to be sure to not allow payment
mode not cofigured

Rename payment_mode into payment_mode_id as it's an ID

Fix: inheriting "invader.payment.service", shopinvader_payment module
should inherit with a new "_name" to avoid changing the behaviours of
base components.

Use dispatch in test instead of calling the method directly

sebastienbeau committed:
[REF] implement check on allowed payment on payable

sebastienbeau committed:
[IMP] Improve test and check acquirer

Check that acquirer match the service with test
Add test on manual payment

sebastienbeau committed:
[REF] start to test payment_sips

Restore payment_mode_id validator
Start adding test
Add demo data

sebastienbeau committed:
[REF] make the code consistent

sebastienbeau committed:
[IMP] improve code, add missing readme

sebastienbeau committed:
[REF] simplify code

add _check_acquirer method in payment_service
add a property for get invader.payment component

lmignon committed:
[IMP] invader.payable: Rename invader_start_payment into invader_set_paymet_mode
The goal is to have a clear sepration between the way the payment process is notified to a payable (events) and the initialization of the payment mode on the payable

lmignon committed:
[IMP] shopinvader_payment_manual: Confirm cart on transaction pending state only for manual payment

lmignon committed:
[FIX] test_shopinvader_payment, test_invader_payment: Restore disabled tests

lmignon committed:
[FIX] Contributors

sebastienbeau reviewed (approved):
\o/

lmignon commented:
/ocabot merge major
